### ISSUE TITLE
Significant Performance Impact of adding Custom Transaction Attributes

### ISSUE THREAD

**Issue by jifeingo** [newrelic/newrelic-dotnet-agent#274]
**Description** When using the Transaction API to add custom parameters to a transaction, customers may experience a significant performance impact.

**Expected Behavior** Should be able to add attributes to a transaction (or span) without a significant performance impact.

**Steps to Reproduce** \* Create simple MVC application with the following controller action ``` public ActionResult Index(int cnt) {

```
for (var i = 0; i < cnt; i++){ var key = Guid.NewGuid().ToString(); switch (i % 4) { case 0: NewRelic.Api.Agent.NewRelic.AddCustomParameter(key, Guid.NewGuid().ToString()); break; case 1: NewRelic.Api.Agent.NewRelic.AddCustomParameter(key, DateTime.Now); break; case 2: NewRelic.Api.Agent.NewRelic.AddCustomParameter(key, i); break; case 3: NewRelic.Api.Agent.NewRelic.AddCustomParameter(key, key); break; }}return View();
```

} ```

Please note that `AddCustomParameter` vs `AddCustomAttribute` behave the same.

- Execute the controller action under load measuring CPU Performance of the web server in three configurations:
  1. No Agent (Agent Disabled)
  2. Without adding any attributes `/home/index?cnt=0`. This would test the overhead of the agent without the API calls.
  3. Adding Attributes `home/index?cnt=50`

**Observed Results** \* 32 workers submitting continuous requests \* 4-VPU, 16GB RAM (Azure Standard B4ms)

Config|0 Attribs|32 Attribs|50 Attribs
-|-|-|-
No Agent - NoOp API Calls still made|50%|50%|50%
Agent v8.32|75%|100%|100% Payload too large messages in log

**Your Environment** \* Windows, .NET Framework 4.7, ASP.NET MVC \* New Relic .NET Agent v8.32 (matching API)

**Additional context** [https://newrelic.zendesk.com/hc/en-us/requests/426384](https://newrelic.zendesk.com/hc/en-us/requests/426384)
[Internal Only - GTSE](https://newrelic.atlassian.net/browse/GTSE-4907)
Aha! Link: https://newrelic.aha.io/features/DOTNET-299

**Comment by ericsampson** [comment]
@jifeingo would the baseline agent-only CPU usage you measured get improved by your fix in #269?

Do you think that this issue (effect of attributes) is specific starting with a certain agent version?

Thanks

**Comment by vuqtran88** [comment]
@ericsampson we have not done high level performance testing for the fix in [#269](https://github.com/newrelic/newrelic-dotnet-agent/issues/269) so it might or might not improve the baseline agent-only CPU usage case.

About this issue, we are still investigating why and when this started to occur.

**Comment by nr-ahemsath** [comment]
Summary: the test scenario described in the original issue description is an unusual use case, using unique keys for each custom attribute added, which causes the agent overhead described.

Note how in the reproduction app described above, a new unique custom attribute key is created for each call to `AddCustomParameter(key, value)`.  Each unique custom attribute key created causes a fair amount of work to be done in the agent, due to the agent's architecture for determining which attributes get sent to which kinds of events (transaction events, error events, span events, transaction traces).  See https://github.com/newrelic/newrelic-dotnet-agent/blob/main/src/Agent/NewRelic/Agent/Core/Attributes/AttributeDefinitionService.cs for an idea of the complexity involved.

Given the common use cases for custom attributes (searching and filtering transactions/events based on the values of custom attributes), it would not seem to make sense to create large numbers of unique custom attribute keys.  The performance issue described above does not apply to the expected use cases for custom attributes.
